### PR TITLE
K8s service bindings

### PR DESCRIPTION
**Pull request by dtillman**
## Description

Initial alpha implementation of the K8s Service bindings support.

### CODE REVIEW

File: src/Configuration/src/Abstractions/PostProcessorConfigurationSource.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        RegisteredProcessors = new List<IConfigurationPostProcessor>();
+    }
+
+    public void RegisterPostProcessor(IConfigurationPostProcessor processor)
+    {
+        if (processor == null)
+        {
+            throw new ArgumentNullException(nameof(processor));
+        }
```
Comment (TimHess) [inline src/Configuration/src/Abstractions/PostProcessorConfigurationSource.cs:25]:
```suggestion
        ArgumentGuard.NotNull(processor);
```

File: src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                        result.Add(fileInfo.Name, ReadContentsAsString(file));
+                    }
+                }
+
+                return result;
+            }
+            catch (Exception)
+            {
+                // Log
+                throw new ArgumentException("${path} is invalid");
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs:228]:
```suggestion
                throw new ArgumentException($"Path '{path}' is invalid");
```
Looks like broken string interpolation, does this look like an improvement?

File: src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    public static readonly string KubernetesBindingsPrefix = "k8s" + ConfigurationPath.KeyDelimiter + "bindings";
+
+    private readonly IDisposable _changeTokenRegistration;
+
+    private readonly ServiceBindingConfigurationSource _source;
+
+    public ServiceBindingConfigurationProvider(ServiceBindingConfigurationSource source)
+        : base(source)
+    {
+        _source = source ?? throw new ArgumentNullException(nameof(source));
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs:31]:
```suggestion
        ArgumentGuard.NotNull(source);
        _source = source;
```

File: src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+            IDirectoryContents contents = fileProvider.GetDirectoryContents("/");
+
+            foreach (IFileInfo element in contents)
+            {
+                if (element != null && element.Exists)
+                {
+                    Bindings.Add(new ServiceBinding(element.PhysicalPath));
+                }
+            }
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationProvider.cs:279]:
```suggestion
            foreach (IFileInfo element in contents.Where(element => element?.Exists == true))
            {
                Bindings.Add(new ServiceBinding(element.PhysicalPath));
            }
```
optional

File: src/Configuration/src/Kubernetes.ServiceBinding/LegacyConnectorsPostProcessors.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+namespace Steeltoe.Configuration.Kubernetes.ServiceBinding;
+
+internal sealed class RabbitMQLegacyConnectorPostProcessor : IConfigurationPostProcessor
+{
+    internal const string BindingTypeKey = "rabbitmq";
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/LegacyConnectorsPostProcessors.cs:9]: do you see a use for these `const` values outside these classes, or should they be private?

File: src/Configuration/src/Kubernetes.ServiceBinding/PostProcessorConfigurationProviderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    public static bool IsSteeltoeBindingsEnabled(this PostProcessorConfigurationProvider provider)
+    {
+        return GetValue(provider, "steeltoe:kubernetes:bindings:enable", false);
+    }
+
+    private static bool GetValue(PostProcessorConfigurationProvider provider, string key, bool defaultValue)
+    {
+        return provider.Source.ParentConfiguration != null ? provider.Source.ParentConfiguration.GetValue(key, defaultValue) : defaultValue;
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/PostProcessorConfigurationProviderExtensions.cs:23]:
```suggestion
        return provider.Source.ParentConfiguration?.GetValue(key, defaultValue) ?? defaultValue;
```

File: src/Configuration/src/Kubernetes.ServiceBinding/PostProcessors.cs [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+            return;
+        }
+
+        configData.Filter(ServiceBindingConfigurationProvider.KubernetesBindingsPrefix, ServiceBindingConfigurationProvider.TypeKey, BindingTypeKey).ForEach(
+            bindingNameKey =>
+            {
+                // Spring -> spring.security.oauth2.client....
+                // Steeltoe -> steeltoe:oauth2:binding-name:....
+                var mapper = new ServiceBindingMapper(configData, bindingNameKey, "steeltoe", "oauth2", ConfigurationPath.GetSectionKey(bindingNameKey));
+                string provider = mapper.BindingProvider;
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/PostProcessors.cs:525]: `provider` is already used on line 512, maybe this should be bindingProvider?

File: src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationSource.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            FileProvider = fileProvider;
+        }
+    }
+
+    public IConfigurationProvider Build(IConfigurationBuilder builder)
+    {
+        if (ParentConfiguration == null)
+        {
+            ParentConfiguration = GetParentConfiguration(builder);
+        }
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationSource.cs:55]:
```suggestion
        ParentConfiguration ??= GetParentConfiguration(builder);
```

File: src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationSource.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    public Predicate<string> IgnoreKeyPredicate { get; set; } = p => false;
+
+    public ServiceBindingConfigurationSource()
+        : this(Environment.GetEnvironmentVariable(ServiceBindingRootDirEnvVariable))
+    {
+    }
+
+    public ServiceBindingConfigurationSource(string serviceBindingRootDirectory)
+    {
+        ServiceBindingRoot = serviceBindingRootDirectory != null ? Path.GetFullPath(serviceBindingRootDirectory) : serviceBindingRootDirectory;
```
Comment (TimHess) [inline src/Configuration/src/Kubernetes.ServiceBinding/ServiceBindingConfigurationSource.cs:34]:
```suggestion
        ServiceBindingRoot = serviceBindingRootDirectory != null ? Path.GetFullPath(serviceBindingRootDirectory) : null;
```

File: src/Configuration/test/Kubernetes.ServiceBinding.Test/ServiceBindingConfigurationProviderTest.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        var builder = new ConfigurationBuilder();
+        builder.Add(new ServiceBindingConfigurationSource(GetEmptyK8SResourcesDirectory()));
+        IConfigurationRoot configuration = builder.Build();
+        Assert.NotNull(configuration);
+        Assert.Throws<InvalidOperationException>(() => configuration.GetRequiredSection("k8s"));
+    }
+
+    [Fact]
+    public void PostProcessors_DisabledbyDefault()
```
Comment (TimHess) [inline src/Configuration/test/Kubernetes.ServiceBinding.Test/ServiceBindingConfigurationProviderTest.cs:108]:
```suggestion
    public void PostProcessors_OffByDefault()
```

**Review by TimHess — CHANGES_REQUESTED** [review]
Mostly minor suggestions, additionally test classes could be `sealed`. Seems nearly there

File: src/Configuration/src/Kubernetes.ServiceBinding/LegacyConnectorsPostProcessors.cs (hunk 5 above)
Comment (dtillman) [reply src/Configuration/src/Kubernetes.ServiceBinding/LegacyConnectorsPostProcessors.cs:9]: Used in unit tests

File: src/Configuration/src/Kubernetes.ServiceBinding/PostProcessors.cs (hunk 7 above)
Comment (dtillman) [reply src/Configuration/src/Kubernetes.ServiceBinding/PostProcessors.cs]: Changed

**Review by TimHess — APPROVED** [review]

**Pull request merged**